core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -255,6 +255,10 @@ class Partition(val topicPartition: TopicPartition,
 
   def isAddingReplica(replicaId: Int): Boolean = assignmentState.isAddingReplica(replicaId)
 
+  // For advancing the HW we assume the largest ISR even if the controller hasn't made the change yet
+  // This set includes the latest ISR (as we learned from LeaderAndIsr) and any replicas from a pending ISR expansion
+  def effectiveInSyncReplicaIds: Set[Int] = inSyncReplicaIds | pendingInSyncReplicaIds
```

Comment:
Since we are now only allowing one in-flight AlterIsr, I changed the semantics of pendingInSyncReplicaIds to be the maximal "effective" ISR. This way we don't need to compute it each time.